### PR TITLE
Optimize Metadata

### PR DESCRIPTION
**Pull request by rennergade**
This PR does a bunch of things concerning storing metadata

1. It implements a log system which logs individual metadata updates to a separate file, which syncs if a former crash is detected at start up
2. It changes the serialization format from JSON to CBOR
3. It adds infrastructure to mmap the logfile and write to it so it updates async

### CODE REVIEW

**Comment by rennergade** [comment]
This is going to fail the CI because it heavily changes infrastructure, but it passes the test cases on my end.

Well have to manually reset the filesystem on the runner once this is merged.

File: src/safeposix/syscalls/fs_calls.rs [hunk 1]
```diff
@@ -177,8 +178,7 @@ impl Cage {
                 } //insert a reference to the file in the parent directory
                 else {unreachable!();}
                 mutmetadata.inodetable.insert(newinodenum, newinode);
-
-                persist_metadata(&mutmetadata);
+                log_metadata(&mutmetadata, newinodenum);
```
Comment (jesings) [inline src/safeposix/syscalls/fs_calls.rs:181]: Probably also need to log parent dir

File: src/safeposix/syscalls/fs_calls.rs [hunk 2]
```diff
@@ -351,8 +350,7 @@ impl Cage {
 
                     } //we don't need a separate unlinked flag, we can just check that refcount is 0
                 }
-                persist_metadata(&mutmetadata);
-
+                log_metadata(&mutmetadata, inodenum);
```
Comment (jesings) [inline src/safeposix/syscalls/fs_calls.rs:353]: log parent dir too

File: src/safeposix/syscalls/fs_calls.rs [hunk 3]
```diff
@@ -1244,7 +1242,7 @@ impl Cage {
                             if dir_inode_obj.linkcount == 2 && dir_inode_obj.refcount == 0 {
                                 //removing the file from the metadata 
                                 mutmetadata.inodetable.remove(&inodenum);
-                                persist_metadata(&mutmetadata);
+                                log_metadata(&mutmetadata, inodenum);
```
Comment (jesings) [inline src/safeposix/syscalls/fs_calls.rs:1245]: log parent dir as well

File: src/safeposix/syscalls/fs_calls.rs [hunk 4]
```diff
@@ -1258,7 +1256,7 @@ impl Cage {
                             if char_inode_obj.linkcount == 0 && char_inode_obj.refcount == 0 {
                                 //removing the file from the metadata 
                                 mutmetadata.inodetable.remove(&inodenum);
-                                persist_metadata(&mutmetadata);
+                                log_metadata(&mutmetadata, inodenum);
```
Comment (jesings) [inline src/safeposix/syscalls/fs_calls.rs:1259]: log parent dir as well

File: src/safeposix/syscalls/fs_calls.rs [hunk 5]
```diff
@@ -1593,14 +1592,15 @@ impl Cage {
                     return syscall_error(Errno::EOPNOTSUPP, "rename", "Cannot move file to another directory");
                 }
                 
-                if let Inode::Dir(parent_dir) = metadata.inodetable.get_mut(&parent_inodenum).unwrap() {
+                let pardir_inodeobj = metadata.inodetable.get_mut(&parent_inodenum).unwrap();
+                if let Inode::Dir(parent_dir) = pardir_inodeobj {
```
Comment (jesings) [inline src/safeposix/syscalls/fs_calls.rs:1596]: I think this is probably left over from some previous architecture you had, this change doesn't really matter one way or another

File: src/tools/fs_utils.rs [hunk 6]
```diff
@@ -162,10 +162,6 @@ fn main() {
             *metadata = FilesystemMetadata::blank_fs_init();
             drop(metadata);
             load_fs_special_files(&utilcage);
-
-            let metadata2 = FS_METADATA.read().unwrap();
-            persist_metadata(&*metadata2);
-            return;
```
Comment (jesings) [inline src/tools/fs_utils.rs:168]: Wait why don't you want us to persist formatting the lindfs?

**Review by jesings — APPROVED** [review]
Awesome changes, I've commented some nitpicks and then some correctness fixes

File: src/interface/file.rs [hunk 7]
```diff
@@ -274,6 +276,175 @@ impl EmulatedFile {
...
+
+        let mut openfiles = OPEN_FILES.lock().unwrap();
+
+        if openfiles.contains(&filename) {
+            panic!("FileInUse");
+        }
+
+        let path: RustPathBuf = [".".to_string(), filename.clone()].iter().collect();
+        let f = OpenOptions::new().read(true).write(true).create(true).open(filename.clone()).unwrap();
+        let absolute_filename = fs::canonicalize(&path)?;
```
[the remark this answers is not in the document]
Comment (jesings) [reply src/interface/file.rs:310]: I mean that you use a library function without importing the function form the library, which we should probably do for clarity

[the remark this answers is not in the document]
Comment (rennergade) [reply src/interface/file.rs:310]: Ah got it. Yeah we did this in emulated file so I just copied. Ill change both.

File: src/interface/file.rs [hunk 8]
```diff
@@ -274,6 +276,175 @@ impl EmulatedFile {
...
+        openfiles.insert(filename.clone());
+
+        let mapsize = usize::pow(2, 20);   
+        // set the file equal to where were mapping the count and the actual map
+        let _newsize = f.set_len((COUNTMAPSIZE + mapsize) as u64).unwrap();
```
[the remark this answers is not in the document]
Comment (rennergade) [reply src/interface/file.rs:315]: We need to do this when mapping new regions. We can't map a file with not enough length. 

What we don't do anymore is constantly tamper with the length after mapping.

**Comment by kapkic** [comment]
Looks good!
> This is going to fail the CI because it heavily changes infrastructure, but it passes the test cases on my end.
If we are going to push changes to infrastructure a bit more regularly in the future, we can modify Actions to lindfs to reformat-reinstall so it wouldn't fail.

**Review by kapkic — APPROVED** [review]

File: src/tools/fs_utils.rs (hunk 6 above)
Comment (rennergade) [reply src/tools/fs_utils.rs:168]: It now falls through to rustfinalize which will persist.

**Comment by rennergade** [comment]
Alright I think I've addressed everything here!

File: src/safeposix/filesystem.rs [hunk 9]
```diff
@@ -123,18 +132,59 @@ pub fn load_fs() {
...
+
     } else {
        *mutmetadata = FilesystemMetadata::blank_fs_init();
        drop(mutmetadata);
+       create_log();
 
        load_fs_special_files(&utilcage);
-
        let metadata = FS_METADATA.read().unwrap();
        persist_metadata(&metadata);
```
[the remark this answers is not in the document]
Comment (jesings) [reply src/safeposix/filesystem.rs:176]: Just wanted to ask again about this?

[the remark this answers is not in the document]
Comment (jesings) [reply src/safeposix/filesystem.rs:176]: Ah probably because you don't want a case to init the blank fs and then restore from log. Noted

**Review by jesings — APPROVED** [review]
Looks good now!

**Pull request merged**